compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -273,55 +397,134 @@ pub fn push_debuginfo_type_name<'tcx>(
 
             output.push_str("enum$<");
             push_item_name(tcx, def.did, true, output);
-            push_type_params(tcx, substs, output, visited);
+            push_generic_params_internal(tcx, substs, output, visited);
 
             let dataful_variant_name = def.variants[*dataful_variant].ident.as_str();
 
             output.push_str(&format!(", {}, {}, {}>", min, max, dataful_variant_name));
         } else {
             output.push_str("enum$<");
             push_item_name(tcx, def.did, true, output);
-            push_type_params(tcx, substs, output, visited);
-            output.push('>');
+            push_generic_params_internal(tcx, substs, output, visited);
+            push_close_angle_bracket(tcx, output);
         }
     }
+}
+
+pub fn push_item_name(tcx: TyCtxt<'tcx>, def_id: DefId, qualified: bool, output: &mut String) {
+    let def_key = tcx.def_key(def_id);
+    if qualified {
+        if let Some(parent) = def_key.parent {
+            push_item_name(tcx, DefId { krate: def_id.krate, index: parent }, true, output);
+            output.push_str("::");
+        }
+    }
+
+    push_unqualified_item_name(tcx, def_id, def_key.disambiguated_data, output);
+}
 
-    fn push_item_name(tcx: TyCtxt<'tcx>, def_id: DefId, qualified: bool, output: &mut String) {
-        if qualified {
+fn push_unqualified_item_name(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    disambiguated_data: DisambiguatedDefPathData,
+    output: &mut String,
+) {
+    let cpp_like_names = tcx.sess.target.is_like_msvc;
+
+    match disambiguated_data.data {
+        DefPathData::CrateRoot => {
             output.push_str(&tcx.crate_name(def_id.krate).as_str());
-            for path_element in tcx.def_path(def_id).data {
-                write!(output, "::{}", path_element.data).unwrap();
+        }
+        DefPathData::Impl => {
+            let self_ty = tcx.type_of(def_id);
+            output.push_str(if cpp_like_names { "impl$<" } else { "<impl " });
+            if let Some(impl_trait_ref) = tcx.impl_trait_ref(def_id) {
+                push_item_name(tcx, impl_trait_ref.def_id, true, output);
+                output.push_str(if cpp_like_names { ", " } else { " for " });
+            }
```

Comment:
Let use the [standard path syntax](https://rust-lang.github.io/rfcs/0132-ufcs.html#path-syntax) here:
- `<SelfType>` for inherent impls, and
- `<SelfType as Trait>` for trait impls.

For MSVC we'd still use `impl$<...>` but note that the order between self-type and trait is reversed.

Reply:
Per our discussion, I'm simplifying this code and we'll address this with #86167